// electric-sql/dist/util/types.js
class SatelliteError extends Error {
  code;
  constructor(code, message) {
    super(message);
    this.code = code;
  }
}
var SatelliteErrorCode = /* @__PURE__ */ ((SatelliteErrorCode2) => {
  SatelliteErrorCode2[SatelliteErrorCode2["CONNECTION_CANCELLED_BY_DISCONNECT"] = 0] = "CONNECTION_CANCELLED_BY_DISCONNECT";
  SatelliteErrorCode2[SatelliteErrorCode2["CONNECTION_FAILED_AFTER_RETRY"] = 1] = "CONNECTION_FAILED_AFTER_RETRY";
  SatelliteErrorCode2[SatelliteErrorCode2["INTERNAL"] = 2] = "INTERNAL";
  SatelliteErrorCode2[SatelliteErrorCode2["TIMEOUT"] = 3] = "TIMEOUT";
  SatelliteErrorCode2[SatelliteErrorCode2["REPLICATION_NOT_STARTED"] = 4] = "REPLICATION_NOT_STARTED";
  SatelliteErrorCode2[SatelliteErrorCode2["REPLICATION_ALREADY_STARTED"] = 5] = "REPLICATION_ALREADY_STARTED";
  SatelliteErrorCode2[SatelliteErrorCode2["UNEXPECTED_STATE"] = 6] = "UNEXPECTED_STATE";
  SatelliteErrorCode2[SatelliteErrorCode2["UNEXPECTED_MESSAGE_TYPE"] = 7] = "UNEXPECTED_MESSAGE_TYPE";
  SatelliteErrorCode2[SatelliteErrorCode2["PROTOCOL_VIOLATION"] = 8] = "PROTOCOL_VIOLATION";
  SatelliteErrorCode2[SatelliteErrorCode2["UNKNOWN_DATA_TYPE"] = 9] = "UNKNOWN_DATA_TYPE";
  SatelliteErrorCode2[SatelliteErrorCode2["SOCKET_ERROR"] = 10] = "SOCKET_ERROR";
  SatelliteErrorCode2[SatelliteErrorCode2["UNRECOGNIZED"] = 11] = "UNRECOGNIZED";
  SatelliteErrorCode2[SatelliteErrorCode2["FATAL_ERROR"] = 12] = "FATAL_ERROR";
  SatelliteErrorCode2[SatelliteErrorCode2["AUTH_ERROR"] = 13] = "AUTH_ERROR";
  SatelliteErrorCode2[SatelliteErrorCode2["AUTH_FAILED"] = 14] = "AUTH_FAILED";
  SatelliteErrorCode2[SatelliteErrorCode2["AUTH_REQUIRED"] = 15] = "AUTH_REQUIRED";
  SatelliteErrorCode2[SatelliteErrorCode2["AUTH_EXPIRED"] = 16] = "AUTH_EXPIRED";
  SatelliteErrorCode2[SatelliteErrorCode2["INVALID_REQUEST"] = 17] = "INVALID_REQUEST";
  SatelliteErrorCode2[SatelliteErrorCode2["PROTO_VSN_MISMATCH"] = 18] = "PROTO_VSN_MISMATCH";
  SatelliteErrorCode2[SatelliteErrorCode2["REPLICATION_FAILED"] = 19] = "REPLICATION_FAILED";
  SatelliteErrorCode2[SatelliteErrorCode2["BEHIND_WINDOW"] = 20] = "BEHIND_WINDOW";
  SatelliteErrorCode2[SatelliteErrorCode2["INVALID_POSITION"] = 21] = "INVALID_POSITION";
  SatelliteErrorCode2[SatelliteErrorCode2["SUBSCRIPTION_NOT_FOUND"] = 22] = "SUBSCRIPTION_NOT_FOUND";
  SatelliteErrorCode2[SatelliteErrorCode2["SUBSCRIPTION_ERROR"] = 23] = "SUBSCRIPTION_ERROR";
  SatelliteErrorCode2[SatelliteErrorCode2["MALFORMED_LSN"] = 24] = "MALFORMED_LSN";
  SatelliteErrorCode2[SatelliteErrorCode2["UNKNOWN_SCHEMA_VSN"] = 25] = "UNKNOWN_SCHEMA_VSN";
  SatelliteErrorCode2[SatelliteErrorCode2["SHAPE_REQUEST_ERROR"] = 26] = "SHAPE_REQUEST_ERROR";
  SatelliteErrorCode2[SatelliteErrorCode2["SUBSCRIPTION_ID_ALREADY_EXISTS"] = 27] = "SUBSCRIPTION_ID_ALREADY_EXISTS";
  SatelliteErrorCode2[SatelliteErrorCode2["SUBSCRIPTION_ALREADY_EXISTS"] = 28] = "SUBSCRIPTION_ALREADY_EXISTS";
  SatelliteErrorCode2[SatelliteErrorCode2["UNEXPECTED_SUBSCRIPTION_STATE"] = 29] = "UNEXPECTED_SUBSCRIPTION_STATE";
  SatelliteErrorCode2[SatelliteErrorCode2["TABLE_NOT_FOUND"] = 30] = "TABLE_NOT_FOUND";
  SatelliteErrorCode2[SatelliteErrorCode2["REFERENTIAL_INTEGRITY_VIOLATION"] = 31] = "REFERENTIAL_INTEGRITY_VIOLATION";
  SatelliteErrorCode2[SatelliteErrorCode2["EMPTY_SHAPE_DEFINITION"] = 32] = "EMPTY_SHAPE_DEFINITION";
  SatelliteErrorCode2[SatelliteErrorCode2["DUPLICATE_TABLE_IN_SHAPE_DEFINITION"] = 33] = "DUPLICATE_TABLE_IN_SHAPE_DEFINITION";
  SatelliteErrorCode2[SatelliteErrorCode2["INVALID_WHERE_CLAUSE_IN_SHAPE_DEFINITION"] = 34] = "INVALID_WHERE_CLAUSE_IN_SHAPE_DEFINITION";
  SatelliteErrorCode2[SatelliteErrorCode2["INVALID_INCLUDE_TREE_IN_SHAPE_DEFINITION"] = 35] = "INVALID_INCLUDE_TREE_IN_SHAPE_DEFINITION";
  SatelliteErrorCode2[SatelliteErrorCode2["SHAPE_DELIVERY_ERROR"] = 36] = "SHAPE_DELIVERY_ERROR";
  SatelliteErrorCode2[SatelliteErrorCode2["SHAPE_SIZE_LIMIT_EXCEEDED"] = 37] = "SHAPE_SIZE_LIMIT_EXCEEDED";
  SatelliteErrorCode2[SatelliteErrorCode2["REPLICATION_TRANSFORM_ERROR"] = 38] = "REPLICATION_TRANSFORM_ERROR";
  return SatelliteErrorCode2;
})(SatelliteErrorCode || {});
var DataChangeType = /* @__PURE__ */ ((DataChangeType2) => {
  DataChangeType2["INSERT"] = "INSERT";
  DataChangeType2["UPDATE"] = "UPDATE";
  DataChangeType2["DELETE"] = "DELETE";
  DataChangeType2["COMPENSATION"] = "COMPENSATION";
  DataChangeType2["GONE"] = "GONE";
  return DataChangeType2;
})(DataChangeType || {});
function isDataChange(change) {
  return "relation" in change;
}
var ReplicationStatus = /* @__PURE__ */ ((ReplicationStatus2) => {
  ReplicationStatus2[ReplicationStatus2["STOPPED"] = 0] = "STOPPED";
  ReplicationStatus2[ReplicationStatus2["STARTING"] = 1] = "STARTING";
  ReplicationStatus2[ReplicationStatus2["STOPPING"] = 2] = "STOPPING";
  ReplicationStatus2[ReplicationStatus2["ACTIVE"] = 3] = "ACTIVE";
  return ReplicationStatus2;
})(ReplicationStatus || {});
export {
  DataChangeType,
  ReplicationStatus,
  SatelliteError,
  SatelliteErrorCode,
  isDataChange
};
