// electric-sql/dist/client/conversions/types.js
var PgBasicType = /* @__PURE__ */ ((PgBasicType2) => {
  PgBasicType2["PG_BOOL"] = "BOOL";
  PgBasicType2["PG_INT"] = "INT";
  PgBasicType2["PG_INT2"] = "INT2";
  PgBasicType2["PG_INT4"] = "INT4";
  PgBasicType2["PG_INT8"] = "INT8";
  PgBasicType2["PG_INTEGER"] = "INTEGER";
  PgBasicType2["PG_REAL"] = "REAL";
  PgBasicType2["PG_FLOAT4"] = "FLOAT4";
  PgBasicType2["PG_FLOAT8"] = "FLOAT8";
  PgBasicType2["PG_TEXT"] = "TEXT";
  PgBasicType2["PG_VARCHAR"] = "VARCHAR";
  PgBasicType2["PG_CHAR"] = "CHAR";
  PgBasicType2["PG_UUID"] = "UUID";
  PgBasicType2["PG_JSON"] = "JSON";
  PgBasicType2["PG_JSONB"] = "JSONB";
  PgBasicType2["PG_BYTEA"] = "BYTEA";
  return PgBasicType2;
})(PgBasicType || {});
var PgDateType = /* @__PURE__ */ ((PgDateType2) => {
  PgDateType2["PG_TIMESTAMP"] = "TIMESTAMP";
  PgDateType2["PG_TIMESTAMPTZ"] = "TIMESTAMPTZ";
  PgDateType2["PG_DATE"] = "DATE";
  PgDateType2["PG_TIME"] = "TIME";
  PgDateType2["PG_TIMETZ"] = "TIMETZ";
  return PgDateType2;
})(PgDateType || {});
function isPgDateType(pgType) {
  return Object.values(PgDateType).includes(pgType);
}
export {
  PgBasicType,
  PgDateType,
  isPgDateType
};
